File: custom_dir/custom_actions/repeat_challenge_n_times.py
import json
import time
from maa.context import Context
from maa.custom_action import CustomAction

from maa.agent.agent_server import AgentServer

@AgentServer.custom_action("RepeatChallengeNTimes")
class RepeatChallengeNTimes(CustomAction):
    def run(self,
            context: Context,
            argv: CustomAction.RunArg) -> bool:
        """
        数字识别验证动作
        :param argv: 运行参数, 包括 start_repeat, expected_number
        :param context: 运行上下文
        :return: 是否执行成功
        """

        # 读取 custom_param 的参数
        # {"start_repeat": true, "expected_number": x}
        json_data = json.loads(argv.custom_action_param)

        # 获取参数
        start_repeat = json_data.get("start_repeat", True)
        expected_number = str(json_data.get("expected_number", 1))


        if not expected_number :
            print("无效的参数：expected_number")
            return False

        print(f"开始点击自动允许次数设置：期望数字 {expected_number}")
        if start_repeat:
            context.run_task("通用_启动设置挑战次数")
            print("启动自动挑战")
        else:
            context.run_task("通用_取消设置挑战次数")
            print("关闭自动挑战")
            return True



        # 识别当前区域的数字
        current_number = self._recognize_number(context)
        print(f"当前设置次数为{current_number}")

        if current_number == expected_number:
            print("当前数字与期望数字相同,完成次数设置")
        else:
            self.input_expected_number(context,expected_number)


        return True

    def _recognize_number(self, context: Context) -> str:
        """
        识别指定区域的数字
        :param context: 运行上下文
        :param
        :return: 识别到的数字
        """

        # 运行识别任务
        img = context.tasker.controller.post_screencap().wait().get()
        result = context.run_recognition("通用_识别挑战次数",img)
        if result and result.best_result:
            try:
                value = str(result.best_result.text)
                return value
            except (ValueError, TypeError):
                return "10"
        else:
            return "10"

    def input_expected_number(self,context: Context,expected_number):
        print("开始点击数字")
        context.run_task("设置挑战次数_点击数字编辑")
        number_list=list(expected_number)
        print(number_list)
        for n in number_list:
            print(f"开始点击数字: {n}")
            context.run_task("设置挑战次数_点击目标数字",{"设置挑战次数_点击目标数字":{"expected":n}})
        print("点击数字完成")
        nu=self._recognize_number(context)
        if nu == expected_number:
            context.run_task("设置挑战次数_点击确定")

    def stop(self) -> None:
        """停止执行的逻辑"""
        print("停止数字验证动作")
        pass

File: agent/actions/repeat_challenge_n_times/repeat_challenge_n_times.go
package repeat_challenge_n_times

import (
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/MaaXYZ/maa-framework-go/v4"
)

type RepeatChallengeNTimes struct{}

type RepeatParams struct {
	StartRepeat    bool `json:"start_repeat"`
	ExpectedNumber int  `json:"expected_number"`
}

// Run 设置挑战次数
func (a *RepeatChallengeNTimes) Run(ctx *maa.Context, arg *maa.CustomActionArg) bool {
	params := RepeatParams{StartRepeat: true, ExpectedNumber: 1}
	if arg.CustomActionParam != "" {
		json.Unmarshal([]byte(arg.CustomActionParam), &params)
	}

	expectedNumber := strconv.Itoa(params.ExpectedNumber)
	if expectedNumber == "" || expectedNumber == "0" {
		fmt.Println("无效的参数：expected_number")
		return false
	}

	fmt.Printf("开始点击自动允许次数设置：期望数字 %s\n", expectedNumber)

	if params.StartRepeat {
		_, _ = ctx.RunTask("通用_启动设置挑战次数", nil)
		fmt.Println("启动自动挑战")
	} else {
		_, _ = ctx.RunTask("通用_取消设置挑战次数", nil)
		fmt.Println("关闭自动挑战")
		return true
	}

	// 识别当前区域的数字
	currentNumber := a.recognizeNumber(ctx)
	fmt.Printf("当前设置次数为%s\n", currentNumber)

	if currentNumber == expectedNumber {
		fmt.Println("当前数字与期望数字相同,完成次数设置")
	} else {
		a.inputExpectedNumber(ctx, expectedNumber)
	}

	return true
}

func (a *RepeatChallengeNTimes) recognizeNumber(ctx *maa.Context) string {
	controller := ctx.GetTasker().GetController()
	if controller == nil {
		return "10"
	}

	controller.PostScreencap().Wait()
	img, err := controller.CacheImage()
	if err != nil {
		return "10"
	}

	result, _ := ctx.RunRecognition("通用_识别挑战次数", img, nil)
	if result != nil && result.Hit {
		return "10"
	}
	return "10"
}

func (a *RepeatChallengeNTimes) inputExpectedNumber(ctx *maa.Context, expectedNumber string) {
	fmt.Println("开始点击数字")
	_, _ = ctx.RunTask("设置挑战次数_点击数字编辑", nil)

	for _, n := range expectedNumber {
		fmt.Printf("开始点击数字: %c\n", n)
		_, _ = ctx.RunTask("设置挑战次数_点击目标数字", map[string]any{
			"设置挑战次数_点击目标数字": map[string]any{"expected": string(n)},
		})
	}

	fmt.Println("点击数字完成")
	nu := a.recognizeNumber(ctx)
	if nu == expectedNumber {
		_, _ = ctx.RunTask("设置挑战次数_点击确定", nil)
	}
}